nowcasting_dataloader/utils/optical_flow.py

```diff
@@ -0,0 +1,219 @@
+"""Functions for computing the optical flow on the fly for satellite images"""
+import logging
+from typing import Optional
+
+import cv2
+import einops
+import numpy as np
+import pandas as pd
+import torch
+import xarray as xr
+from nowcasting_dataset.dataset.batch import Batch
+
+_LOG = logging.getLogger("nowcasting_dataset")
+
+
+def compute_optical_flow_for_batch(
+    batch: Batch, final_image_size_pixels: Optional[int] = None
+) -> torch.Tensor:
+    """
+    Computes the optical flow for satellite images in the batch
+
+    Assumes metadata is also in Batch, for getting t0
+
+    Args:
+        batch: Batch containing at least metadata and satellite data
+
+    Returns:
+        Tensor containing the Optical Flow predictions
+    """
+
+    assert (
+        batch.satellite is not None
+    ), "Satellite data does not exist in batch, required for optical flow"
+    assert batch.metadata is not None, "Metadata does not exist in batch, required for optical flow"
+
+    if final_image_size_pixels is None:
+        final_image_size_pixels = len(batch.satellite.x_index)
+
+    # Only do optical flow for satellite data
+    optical_flow_predictions = []
+    for i in range(batch.batch_size):
```

Comment:
could this loop be over diffetn threads? i.e in parrellle?

Reply:
It could, but the idea being we'll be running as many dataloader threads as we can, and so then parallelizing this would just cause them to contend for the same cpu cores, possibly slowing down the whole thing. So I would prefer the dataloader processes to stay single threaded if possible.

Reply:
And instead run 16 dataloader threads, loading 16 batches in parallel instead